### ISSUE TITLE
Investigate if posing new Fluid content in teams can be optimized

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#5538]
From https://dev.azure.com/office/OC/_workitems/edit/4899481?src=WorkItemMention&src-action=artifact_link

Scenario:
- User creates a Fluid component to send in a Teams chat
- User adds some content to the component
- User presses send
Result:
- User that created the component show up in loading state when it is inserted into the chat rather than an instant hydration.

The key expectation here - initial loading of same container to happen without network calls.
I.e. client to cache enough information to make second render instant.

There is no good way to accomplish that today.
The best we could do is
- cache initial summary that file was created from and convert it to snapshot format (or find other way to load from it directly)
- cache all the ops up to some point.
- when loading, load from snapshot and play ops, instead of going over network for latest snapshot + ops.

Might be simpler, but lesser gain to the user:
- start fetching snapshot from storage sooner, on user click send, outside of original iframe, and block loading of new container until this process is done. It only helps if all the other things (creation of new iframe, etc.) could happen in parallel (i.e. not be blocked by pending fetch) and only fetching for snapshot would block and wait for results from pre-fetch.

**Comment by vladsud** [comment]
@jatgarg, any chance you have time to help me here in next 5-7 business days?
The core of the work here is to convert payload we get from Container when creating new file (ISummaryTree) to a snapshot (IOdspSnapshot), such that we can push it into cache and could reuse on next boot.

**Comment by jatgarg** [comment]
@vladsud yes sure. So should I write an api say in odsp-driver to convert the snapshot?

**Comment by vladsud** [comment]
PR referenced above is big part of a solution. What's missing is adding end-to-end caching and testing

**Comment by vladsud** [comment]
I have not had time to attack it in June, need some help :)
Si - I'd appreciate if you can take care of it.
Jatin - I'd appreciate if you could guide see on actual details here.

Here are two parallel streams of work that we can be doing here:
1. Convert "create new file" summary to a snapshot and cache it. Jatin did preliminary work here, so adding caching is likely rather trivial here. Biggest concern here - testability and correctness. If something goes wrong, it may have rather big impact on user scenario. So we need to think how we will ensure this flow is tested across as many flows as possible (UT in driver, UT / end-to-end tests in runtime, make sure it's included in stress tests, maybe other places).
2. We can cover 95% of that scenario and improve other scenarios simply by fetching snapshot from storage and caching it. This can be more generic than just on create new. For example, we already have prefetchLatestSnapshot() that does all of that - we can call it on specific triggers, like
   - creation of a file
   - When CacheOpsLimitHit event is hit due to too many ops on top of prior cached snapshot. Right now we delete all cached data in such case forcing next boot to go to storage, but we can also simple wire this flow to trigger prefetchLatestSnapshot() can continue ops caching. This is likely more involved (as requires partial ops deleting up to new snapshot), but something to look at (and consider percentage of sessions that run into this case - if it's relatively high, maybe separate issue is warranted to track it)

**Comment by jatgarg** [comment]
@vladsud So as I can understand in teams when a user creates a fluid content in personal textBox is a creation case and then when user hits the send button, the container is attached and then loaded back again in the chat window?
If so, should be consider just dehydrating detached container on send button, rehydrating it back in main chat window and then letting it attach in the background or maybe it is not possible because whatever is in main chat  window should always exists on server? Just a thought so putting it out there?

**Comment by vladsud** [comment]
I do not think we want to go that far, yet :) This would be rather big change for hosting app, as today they may block user from moving away (from given window / chat) while waiting for initial post to finish, but do not care about loaded container.
I think long term we should assume that process of file creation may happen outside of iframe and controlled by host, while containers are more like views that can come and go at any point in time. So fast rendering should not be tied to how long it takes to post a file or where it happens (and lifetime of original container). So the task here is smaller in scope - when attachment process starts, simply cache the snapshot for another container instance to be able to rehydrate from it, if that's possible. If it's not possible, I'd rather wait for creation to finish and invoke prefetchLatestSnapshot() right away. Latter only hosts because all clients create files too early in the process, before user clicks send.

**Comment by jatgarg** [comment]
@chensixx So for now we can implement the first approach where we cache the "CreateNew" snapshot and ops in host persistent cache and then load from that on next Container load as it involves 0 extra network calls whereas the "Prefetching snapshot" involves making network calls.
Steps:
I think for now we can implement it under a policy flag and we will that that policy option in driver factory 

> HostStoragePolicy

.
1.) I implemented the function to convert "createNew" summary to IOdspSnapshot which is the format we fetch from server. However we currently don't use it cache the summary. So you need to add this support of converting and caching that while we create the file in createNew.ts in Odsp driver. This snapshot just contains blobs and trees and no ops as we don't have any ops yet. You can see that in implementation.
2.) Odsp driver already have capability to cache ops when they are emitted from container, so you don't need to do add anything extra for caching of ops in host cache.
3.) Now once both are cached, they will just be fetched from cache during the load time. So imo no extra work will be required there.

Testing:
1.) We do have some unit tests in place for converting "createNew" summary format to IOdspSnapshot but once you cache, you may want to add more unit tests for the same.
2.) I think for this we can add support in "Webpack-fluid-loader" to test the end-to-end flow and see that the contents are actually fetched from cache and no network call is made for the same. Look at the implementation to test "prefetchSnapshot" api in webpack-fluid-loader. Here: https://github.com/microsoft/FluidFramework/blob/c32fff7c6214a8562bb6cf416eed94a6eb502a4f/packages/tools/webpack-fluid-loader/src/loader.ts#L249

**Comment by vladsud** [comment]
@jatgarg, I believe you implemented conversion for # 1 earlier, but it's unused. So the work is mostly around caching it and validation

**Comment by jatgarg** [comment]
Yes. I forogt to paste the link of the api to Si.: https://github.com/microsoft/FluidFramework/blob/4788e9eae17d986947b5e555c935a384fd4da7ca/packages/drivers/odsp-driver/src/createNewUtils.ts#L14

**Comment by jatgarg** [comment]
Reopening so that you can enable default flag after adding test. After that you can close it.
